### DISCUSSION TITLE
(Experiment) Reversing the breath layer

### DISCUSSION THREAD

**Discussion by luttje** [CapsAdmin/pac3#1071] [Help]
**The idea:** Use the breathing layer information from [this smd file](https://github.com/robotboy655/gmod-animations/blob/master/GMod/breath.smd) to generate a Custom Animation that animates each bone in the exact opposite angle and translation. Synchronize the custom animation with the breathing animation using [Entity:SetCycle](https://wiki.facepunch.com/gmod/Entity:SetCycle)

**The challenges:**
1. I assume that SetCycle will be capable of helping synchronize with the breathing layer. No idea if it will.
2. The smd uses positions and angles for all bones relative to the origin of the model, whilst pac wants each position relative to the parent of that bone. [valve docs](https://developer.valvesoftware.com/wiki/Studiomdl_Data#Syntax_2)
3. pac works with up, right, forward, whilst the smd works with X, Y and Z?
4. Again I don't know if the above is possible.

As an experiment I started just writing some code, but honestly I haven't had any success. In lieu of results I've made it fetch and parse the smd from github: luttje/pac3@9a6f2bbe58ca3d16164d3fe1bd65fda089ee2038

I'm writing this discussion instead of a PR since it's too raw of a project. As well as to get some suggestions going. My brain and eyes are hurting too much to try and solve this for today 😅

**Comment by CapsAdmin** [comment]
Maybe I don't understand but can't we just disable the breathing animation? I think it's https://github.com/CapsAdmin/pac3/blob/develop/lua/pac3/editor/client/view.lua#L459-L470

**Reply by luttje** [reply]
After a bit more trying I can't find a way to disable breathing using anything simple.

I think I really have to go about the ugly workaround of adding a custom animation that reverses the effect. Though I can't seem to wrap my head around matrices and how to localize them.

I've made a bit of progress here luttje/pac3@565a7fd1793de857d60984afd1f43c43fd464d9e, though I'm running into the problem that I have no idea how to apply the SMD animation (relative to origin) to a bone (relative to it's parent).

In the code in that branch I've tried to do:
1. Create bone part in the UI and use the tool I'm working on Tools > 'apply reverse breath animation to bone'
2. That tool parses the SMD, creates a custom animation and stores the **manipulation data** (angle and position) for that bone unchanged (only changing angle to be degrees instead of radians)
3. The tool marks the **manipulation data** as "relative_to_world" (named kind of weirdly, relative to origin of model would make more sense)
4. On animating I check if relative_to_world is set for any bones in the frame. 
5. If it is then the **original bone data** (angle and position) is gotten as a matrix relative to the world
6. I use the  **manipulation data** and LocalToWorld(Angles) to create a matrix of the bone manipulation data.
7. I subtract the  **original bone data** from the **world manipulation data**, because in my head that would make the manipulation data relative to the original bone position, instead of the origin of the model.

Though as you may be able to see in the fork I'm having minimal success. I feel like I'm trying to 'rush to a solution' without thinking and I can't seem to stop myself to just take a step back 😭😅

**Comment by luttje** [comment]
> Maybe I don't understand but can't we just disable the breathing animation? I think it's https://github.com/CapsAdmin/pac3/blob/develop/lua/pac3/editor/client/view.lua#L459-L470

I think you understand fine, though that is the first I'm hearing of this possibility and of seeing any code like that.

Searching online and in discords (pac and gmod) revealed no useful results for me yesterday.

I'll play around with some of the code you sent in a bit, that would make this so much easier.

**Comment by luttje** [comment]
I double checked and sadly this `breathing` is not a pose parameter that's available. I tested these default models (I assume the rest have the same problem): `models/player/Group01/male_02.mdl` and `models/player/mossman.mdl`

Checking [the `qc` here](https://github.com/robotboy655/gmod-animations/blob/1ec11d7d92f23fe1359319421301ea0f97eb24d5/gm_sequences.qci#L56) confirms that there's no $poseparameter for this breathing layer sadly

**Reply by luttje** [reply]
The layer does show up if I check with `ply:LookupSequence("gmod_breath_layer")`, but none of the functions listed here help me stop the breathing animation: https://wiki.facepunch.com/gmod/BaseAnimatingOverlay

I'm beginning to wonder if these functions are ~not just bugged~ only meant for animations added with Entity:AddLayeredSequence and dont work for animations compiled into models

**Comment by luttje** [comment]
Here's what I have now and the code that I believe to be incorrect. The orientations are all wrong. Once I fix those we can actually start to see if those breathing animations are being enhanced/reduced.

(I hope you don't mind me just rubber-ducking and venting here, I'm still hoping for some magical insight or suggestion 😅)

PS: If I fix this I think we could convert any existing smd animation to a custom animation in pac3

![weird head roll](https://user-images.githubusercontent.com/2738114/118143439-a329d600-b40b-11eb-865a-27a2bfb6f919.gif)

The code that's giving me nightmares: https://github.com/luttje/pac3/blob/tool-populate-with-reverse-breath/lua/pac3/libraries/animations.lua#L210-L255

I believe I'm using a wrong matrix in converting/translating somewhere. Causing the weird head roll which is sometimes way too much rotation, other times in the wrong direction
